numpy/core/src/multiarray/ctors.c

```diff
@@ -3595,9 +3595,28 @@ array_from_text(PyArray_Descr *dtype, npy_intp num, char *sep, size_t *nread,
     NPY_BEGIN_ALLOW_THREADS;
     totalbytes = bytes = size * dtype->elsize;
     dptr = PyArray_DATA(r);
-    for (i= 0; num < 0 || i < num; i++) {
+    for (i = 0; num < 0 || i < num; i++) {
         if (next(&stream, dptr, dtype, stream_data) < 0) {
-            /* EOF */
+            /*
+             * Normally we should have reached the EOF,
+             * but it could be unmatched data as well, which should error.
+             */
+            int isfile = (stream_data == NULL);
+
+            if (isfile) {
+                if (feof(stream) != EOF) {
+                    if (ftell(stream) == 0) {
+                        /* The file should be empty */
+                        break;
+                    }
```

Comment:
What happens without this if?

Reply:
If you do `np.fromfile(empty_file)` it would raise an error (it did trip a test). But will look if this is not better slightly reorganized.

Reply:
This was all pretty wrong to be honest. Will work on a fixup, too much band-aid thinking instead of digging into the implementation.

Reply:
I think I finished through with the rabbit hole, but should probably pass once more and add tests (only started with those).